src/rust/session/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+//! This crate provides buffered TCP sessions with or without TLS which can be
+//! used with [`::mio`]. TLS/SSL is provided by BoringSSL with the [`::boring`]
+//! crate.
+
+#[macro_use]
+extern crate rustcommon_fastmetrics;
+
+mod buffer;
+mod stream;
+mod tcp_stream;
+
+use std::borrow::Borrow;
+use std::io::BufRead;
+use std::io::{ErrorKind, Read, Write};
+use std::net::SocketAddr;
+
+use boring::ssl::{MidHandshakeSslStream, SslStream};
+use bytes::Buf;
+use common::traits::ExtendFromSlice;
+use metrics::Stat;
+use mio::event::Source;
+use mio::{Interest, Poll, Token};
+
+use buffer::Buffer;
+use stream::Stream;
+
+pub use tcp_stream::TcpStream;
+
+// TODO(bmartin): implement connect/reconnect so we can use this in clients too.
+/// The core `Session` type which represents a TCP stream (with or without TLS),
+/// the session buffer, the mio [`::mio::Token`],
+pub struct Session {
+    token: Token,
+    stream: Stream,
+    read_buffer: Buffer,
+    write_buffer: Buffer,
+    tmp_buffer: Box<[u8]>,
+}
+
+impl Session {
+    /// Create a new `Session` with  representing a plain `TcpStream` with
+    /// internal buffers which can hold up to capacity bytes without
+    /// reallocating.
+    pub fn plain_with_capacity(stream: TcpStream, capacity: usize) -> Self {
+        Self::new(Stream::plain(stream), capacity)
+    }
+
+    /// Create a new `Session` representing a negotiated `SslStream`
+    pub fn tls_with_capacity(stream: SslStream<TcpStream>, capacity: usize) -> Self {
+        Self::new(Stream::tls(stream), capacity)
+    }
+
+    /// Create a new `Session` representing a `MidHandshakeSslStream`
+    pub fn handshaking_with_capacity(
+        stream: MidHandshakeSslStream<TcpStream>,
+        capacity: usize,
+    ) -> Self {
+        Self::new(Stream::handshaking(stream), capacity)
+    }
+
+    /// Create a new `Session`
+    fn new(stream: Stream, capacity: usize) -> Self {
+        increment_counter!(&Stat::TcpAccept);
+        let mut tmp_buffer = vec![0; capacity];
+        tmp_buffer.resize(capacity, 0);
```

Comment:
what's the purpose of this resize considering the line above?

Reply:
Note to self to re-run experiments about allocation to ensure mental model is correct. Feel like this was defensive to avoid some casually observed weird behavior. Will add a notes file or common function with comments if I validate previous finding or fix this code otherwise. Might make sense to use another pattern here if this explicit behavior is actually required.

Reply:
I've changed the pattern to the `with_capacity() + resize(..)` pattern I use elsewhere. It looks like `vec![0; capacity]` results in a lazy allocation, so we'll either need to use the pattern I'm using elsewhere or use the macro initialization + read through the vector.